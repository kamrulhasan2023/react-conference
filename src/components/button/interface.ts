import React from 'react';
import { IconNames } from './../icons/interface';

export const buttonSizes = ['lg', 'md', 'sm', 'xs'] as const;
type Size = (typeof buttonSizes)[number];

export interface IFillButton {
  label: string;
  colorScheme: ColorScheme;
  size: Size;
  clickHandler: (
    e: React.MouseEvent<HTMLElement> | React.FormEvent<HTMLFormElement>
  ) => void;
  hoverColorScheme?: ColorScheme;
  iconName?: IconNames;
  iconIsLeft?: boolean;
  isDisabled?: boolean;
  variant?: 'fill';
  type?: 'button';
}

export interface IFillSubmitButton
  extends Omit<IFillButton, 'clickHandler' | 'type'> {
  type: 'submit';
}

export interface IFillRedirectButton
  extends Omit<IFillButton, 'clickHandler' | 'isDisabled'> {
  href?: string;
  isOpenNewTab: boolean;
}

export interface IOutlineButton
  extends Omit<IFillButton, 'colorScheme' | 'hoverColorScheme' | 'variant'> {
  colorScheme: OutlineColorScheme;
  hoverColorScheme: OutlineColorScheme;
  variant: 'outline';
}

export interface IOutlineSubmitButton
  extends Omit<IOutlineButton, 'clickHandler' | 'type'> {
  type: 'submit';
}

export interface IOutlineRedirectButton
  extends Omit<IOutlineButton, 'clickHandler' | 'isDisabled'> {
  href?: string;
  isOpenNewTab: boolean;
}

export type IBaseButton =
  | Omit<IOutlineButton, 'clickHandler' | 'type'>
  | Omit<IFillButton, 'clickHandler' | 'type'>;

export const buttonColorSchemes = [
  'bg-primary-text-white',
  'bg-success-text-white',
  'bg-black-text-white',
  'bg-white-text-black',
] as const;
type ColorScheme = (typeof buttonColorSchemes)[number];

export const outlineButtonColorSchemes = [
  'bg-transparent-text-primary',
  'bg-transparent-text-success',
] as const;
type OutlineColorScheme = (typeof outlineButtonColorSchemes)[number];
