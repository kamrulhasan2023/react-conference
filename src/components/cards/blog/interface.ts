import { IDimension } from 'src/components/custom-image/interface';

export interface IBlogCard {
  href: string;
  image: {
    src: string;
    alt: string;
    lqip: string;
  };
  title: string;
  description: string;
  tags: {
    bgColor: { hex: string };
    fontColor: { hex: string };
    title: string;
  }[];
  isFeatured: boolean;
}

export const BLOG_IMAGE_DIMENSION: IDimension = {
  width: 960,
  height: 720,
};
