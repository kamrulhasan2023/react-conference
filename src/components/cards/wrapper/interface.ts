export interface ICardWrapper {
  cardShadow: ICardShadow;
  hoverCardShadow?: ICardShadow;
  cardBorderRadius: ICardBorderRadius;
}

export const cardWrapperShadow = ['shadow-one', 'shadow-none'] as const;
type ICardShadow = (typeof cardWrapperShadow)[number];

export const cardWrapperBorderRadius = [
  'radius-one',
  'radius-two',
  'radius-none',
] as const;
type ICardBorderRadius = (typeof cardWrapperBorderRadius)[number];
