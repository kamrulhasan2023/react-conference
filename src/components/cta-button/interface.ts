import { ILabelLink } from '@components/custom-link/interface';

export interface ICtaButton extends ILabelLink {
  colorScheme: ColorScheme;
  hoverColorScheme?: ColorScheme;
}

export const ctaButtonColorSchemes = ['primary', 'success'] as const;
type ColorScheme = (typeof ctaButtonColorSchemes)[number];
