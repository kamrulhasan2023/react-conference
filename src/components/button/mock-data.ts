import {
  IFillButton,
  IFillRedirectButton,
  IOutlineButton,
  IOutlineRedirectButton,
} from './interface';

export const buttonProps: IFillButton = {
  label: 'Button',
  colorScheme: 'bg-primary-text-white',
  hoverColorScheme: 'bg-success-text-white',
  size: 'lg',
  clickHandler: () => {},
  variant: 'fill',
};

export const redirectButtonProps: IFillRedirectButton = {
  href: 'https://www.google.com/',
  isOpenNewTab: true,
  label: 'Redirect Button',
  colorScheme: 'bg-primary-text-white',
  hoverColorScheme: 'bg-success-text-white',
  size: 'lg',
  variant: 'fill',
};

export const outlineButtonProps: IOutlineButton = {
  label: 'Outline Button',
  colorScheme: 'bg-transparent-text-primary',
  hoverColorScheme: 'bg-transparent-text-success',
  size: 'lg',
  clickHandler: () => {},
  variant: 'outline',
};

export const redirectOutlineButtonProps: IOutlineRedirectButton = {
  href: 'https://www.google.com/',
  isOpenNewTab: true,
  label: 'Outline Redirect Button',
  colorScheme: 'bg-transparent-text-primary',
  hoverColorScheme: 'bg-transparent-text-success',
  size: 'lg',
  variant: 'outline',
};
