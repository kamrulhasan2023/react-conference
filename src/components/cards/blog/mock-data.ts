import { IBlogCard } from './interface';

export const blogCardProps: IBlogCard = {
  href: 'https://www.lemonhive.com/',
  image: {
    src: 'https://res.cloudinary.com/dktw59lc2/image/upload/v1669267967/BrightonSEO/Carousel/image-2_yph8q6.jpg',
    alt: 'dummy image 2',
    lqip: 'https://res.cloudinary.com/dktw59lc2/image/upload/v1669267967/BrightonSEO/Carousel/image-2_yph8q6.jpg',
  },
  title: 'Headless Website Development',
  description:
    'Headless solutions built with Sanity.io, Strapi, Prismic, Contentful and others with front-end frameworks such as Next.js, Gatsby, & Nuxt.',
  tags: [
    { fontColor: { hex: 'black' }, bgColor: { hex: 'red' }, title: 'Reactjs' },
    { fontColor: { hex: 'black' }, bgColor: { hex: 'red' }, title: 'Nextjs' },
  ],
  isFeatured: false,
};
