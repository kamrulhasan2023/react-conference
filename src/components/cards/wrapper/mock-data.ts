import { ICardWrapper } from './interface';

export const cardWrapperProps: ICardWrapper = {
  cardBorderRadius: 'radius-none',
  cardShadow: 'shadow-one',
};
