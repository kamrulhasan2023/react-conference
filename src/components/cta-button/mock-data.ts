import { ICtaButton } from './interface';

export const ctaButtonProps: ICtaButton = {
  label: 'Cta button',
  href: '/',
  colorScheme: 'primary',
  isOpenNewTab: false,
};
