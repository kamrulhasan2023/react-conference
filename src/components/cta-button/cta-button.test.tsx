import { render } from '@testing-library/react';
import CtaButton from '.';
import { ctaButtonProps } from './mock-data';
import { ctaButtonColorSchemes } from './interface';

it('renders cta button correctly', () => {
  const { getByRole } = render(<CtaButton {...ctaButtonProps} />);
  const element = getByRole('link');
  expect(element).toBeInTheDocument();
  expect(element).toMatchSnapshot();
});

describe.each([true, false])(
  'Given open new tab configurations',
  (isOpenNewTab) => {
    it(`Expect cta button to contain href and ${
      isOpenNewTab ? 'target=_blank' : 'target=_self'
    }`, () => {
      const props = {
        ...ctaButtonProps,
        href: 'https://www.google.com/',
        isOpenNewTab: isOpenNewTab,
      };
      const { getByRole } = render(<CtaButton {...props} />);
      const link = getByRole('link');

      expect(link).toHaveAttribute('href', 'https://www.google.com/');

      if (isOpenNewTab) {
        expect(link).toHaveAttribute('target', '_blank');
      } else {
        expect(link).toHaveAttribute('target', '_self');
      }
    });
  }
);

describe.each(ctaButtonColorSchemes)('Given color scheme %s', (colorScheme) => {
  it(`Expect cta button to have class ${colorScheme}`, () => {
    const props = { ...ctaButtonProps, colorScheme: colorScheme };
    const { getByRole } = render(<CtaButton {...props} />);
    const button = getByRole('link');
    expect(button).toHaveClass(colorScheme);
  });
});
