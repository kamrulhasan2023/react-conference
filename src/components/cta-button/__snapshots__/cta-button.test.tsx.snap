// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders cta button correctly 1`] = `
<a
  class="relative z-10 overflow-hidden inline-flex items-center justify-center underline transition-colors cta-button primary"
  href="/"
  target="_self"
>
  Cta button
</a>
`;
